### PR TITLE
feat(treasury): add swap functionalities through osmosis 

### PR DESCRIPTION
**Pull request by manu0466**
## Overview

This PR updates the treasury and staking contracts, allowing the staking contract to send protocol fees to the treasury and enabling a `trader` account to perform swap operations, helping the treasury hedge against asset price fluctuations.

To prevent the `trader` from converting the treasury's assets into unwanted ones, a list of swap routes can be defined. This forces the `trader` to use only a predefined set of pools and only in a predefined direction. For example, if the Osmosis pool 1 has the OSMO/USDC pair, we can enforce that the `trader` can only swap OSMO for USDC, but not the other way around.

closes: MILKTIA-35

## What changes have been made in this PR?

- [ ] Add swap functionalities to the treasury contract
- [ ] Allow the staking contract to send the protocol fees to the treasury
- [ ] Fix lint errors reported by `cargo clippy` 

## Checklist

---

- [ ] Appropriate labels applied
- [ ] Targeted PR against correct branch
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote unit and integration
- [ ] Updated relevant documentation

### CODE REVIEW

File: contracts/staking/src/execute.rs [hunk 1]
```diff
@@ -913,16 +928,15 @@
...
-            .unwrap_or(vec![])
+            .as_deref()
+            .unwrap_or_default()
             .iter()
             .any(|v| *v == sender)
-        {
-            return Err(ContractError::Unauthorized { sender });
-        }
+    {
+        return Err(ContractError::Unauthorized { sender });
```
[the remark this answers is not in the document]
Comment (manu0466) [reply contracts/staking/src/execute.rs:939]: This logic is used only to enable the circuit breaker

File: contracts/staking/src/execute.rs [hunk 2]
```diff
@@ -818,24 +820,37 @@
...
 
     // update the accounting of tokens
-    state.total_native_token += amount_after_fees.clone();
-    state.total_reward_amount += amount.clone();
-    state.total_fees += fee;
+    state.total_native_token += amount_after_fees;
+    state.total_reward_amount += amount;
+    if !config.send_fees_to_treasury {
+        state.total_fees += fee;
+    }
```
[the remark this answers is not in the document]
Comment (manu0466) [reply contracts/staking/src/execute.rs:827]: The fee computation logic is used only when receiving the rewards for compaunding. IMO we don't need to move it to a dedicated function.

File: contracts/staking/src/tests/reward_tests.rs [hunk 3]
```diff
@@ -1,101 +1,170 @@
...
+            memo: format!("{{\"ibc_callback\":\"{}\"}}", env.contract.address),
+        })
+    );
+
+    let state = STATE.load(&deps.storage).unwrap();
+
+    assert_eq!(state.total_reward_amount, Uint128::from(100u128));
+    assert_eq!(state.total_native_token, Uint128::from(100_090u128));
+    assert_eq!(state.total_fees, Uint128::from(10u128));
+}
```
[the remark this answers is not in the document]
Comment (manu0466) [reply contracts/staking/src/tests/reward_tests.rs:99]: There is already the `receive_rewards_and_send_fees_to_treasury` to test that.

File: contracts/staking/src/contract.rs [hunk 4]
```diff
@@ -317,7 +321,9 @@ pub fn migrate(mut deps: DepsMut, env: Env, msg: MigrateMsg) -> Result<Response,
     }
 
     let migration_response = match msg {
-        MigrateMsg::V0_4_18ToV0_4_19 {} => migrations::v0_4_19::migrate(deps.branch(), env)?,
+        MigrateMsg::V0_4_18ToV0_4_19 {
```
Comment (rustcandy) [inline contracts/staking/src/contract.rs:324]: maybe 4.19 to 4.20?

File: contracts/staking/src/msg.rs [hunk 5]
```diff
@@ -194,7 +199,7 @@ pub enum QueryMsg {
 
 #[cw_serde]
 pub enum MigrateMsg {
-    V0_4_18ToV0_4_19 {},
+    V0_4_18ToV0_4_20 { send_fees_to_treasury: bool },
```
Comment (hallazzang) [inline contracts/staking/src/msg.rs:202]: Shouldn't this be `V0_4_19ToV0_4_20`?

**Review by hallazzang — CHANGES_REQUESTED** [review]
Overall looks good to me.

File: contracts/staking/src/msg.rs (hunk 5 above)
Comment (manu0466) [reply contracts/staking/src/msg.rs:202]: The version `0.4.19` has never been deployed so we are going to migrate from `0.4.18` to `0.4.20`

**Review by rustcandy — APPROVED** [review]
LGTM

**Review by hallazzang — APPROVED** [review]
LGTM!

**Pull request merged**